nx/lib/nx/random.ex

```diff
@@ -0,0 +1,419 @@
+      |> Nx.as_type({:u, 32})
+
+    state = {xs, ks, rotations}
+
+    {_, {nxs, _, _}} =
+      while {x = 0, state}, Nx.less(x, 5) do
+        {x + 1, rolled_loop_step(x, state)}
+      end
+
+    nxs
+  end
+
+  defnp apply_round(xs, rot) do
+    y1 = Nx.sum(xs, axes: [0])
+
+    y2 =
+      rotate_left(xs[1], rot)
+      |> Nx.bitwise_xor(y1)
+
+    # losing precision on purpose due to upcasts
+    Nx.stack([y1, y2])
+    |> Nx.as_type({:u, 32})
+  end
+
+  defnp rolled_loop_step(i, {xs, ks, rs}) do
+    k1 = ks[0]
+    k2 = ks[1]
+    k3 = ks[2]
+    r1 = rs[0]
+    r2 = rs[1]
+
+    {_, xs, _} =
+      while {x = 0, xs, rs}, x < 4 do
+        {x + 1, apply_round(xs, rs[0][x]), rs}
+      end
+
+    xs1 = k1 + xs[0]
+
+    xs2 = k2 + i + 1 + xs[1]
+
+    new_xs =
+      Nx.stack([xs1, xs2])
+      |> Nx.as_type({:u, 32})
+
+    new_ks = Nx.stack([k2, k3, k1])
+    new_rs = Nx.stack([r2, r1])
+
+    {new_xs, new_ks, new_rs}
+  end
+
+  defnp rotate_left(x, rot) do
+    x <<< rot ||| x >>> (@nbits - rot)
+  end
+
+  defnp float_info(type) do
+    case type do
+      {:bf, 16} -> [exp: 8, mantissa: 7]
+      {:f, 16} -> [exp: 6, mantissa: 10]
+      {:f, 32} -> [exp: 8, mantissa: 23]
+      {:f, 64} -> [exp: 11, mantissa: 52]
+    end
+  end
+
+  @doc """
+  Sample uniform random integer values in [min_val, max_val).
+  
+  ## Options
+  
+    * `:type` - the integer type for the returned tensor
+    * `:shape` - shape of the returned tensor
+  
+  
+  """
+  defn randint(key, min_val, max_val, opts \\ []) do
+    opts = keyword!(opts, shape: {1}, type: {:s, 32})
+    assert_key(key)
+
+    shape = opts[:shape]
+    type = {_, nbits} = normalize(opts[:type])
+
+    case type do
+      {:u, _} ->
+        :ok
+
+      {:s, _} ->
+        :ok
+
+      _ ->
+        raise ArgumentError,
+              "expected integer type, got type #{inspect(type)}"
+    end
+
+    keys = split(key)
+
+    min_val = Nx.broadcast(min_val, shape)
+    max_val = Nx.broadcast(max_val, shape)
+
+    higher_bits = random_bits(keys[0], shape: shape, bit_width: nbits)
+    lower_bits = random_bits(keys[1], shape: shape, bit_width: nbits)
+    span = max_val - min_val
+
+    multiplier =
+      Nx.power(2, Nx.quotient(nbits, 2))
+      |> Nx.remainder(span)
+      |> Nx.power(2)
+      |> Nx.remainder(span)
+
+    offset =
+      higher_bits
+      |> Nx.remainder(span)
+      |> Nx.multiply(multiplier)
+      |> Nx.add(Nx.remainder(lower_bits, span))
+      |> Nx.remainder(span)
+
+    Nx.as_type(min_val + offset, type)
+  end
+
+  @doc """
+  Sample uniform float values in [min_val, max_val).
+  
+  ## Options
+  
+    * `:type` - a float type for the returned tensor
+  
+    * `:shape` - shape of the returned tensor
+  
+    * `:min_val` - minimum value, default is 0.0
+  
+    * `:max_val` - maximum value, default is 1.0
+  
+  """
+  defn uniform(key, opts \\ []) do
+    opts = keyword!(opts, shape: {1}, type: {:f, 32}, min_val: 0.0, max_val: 1.0)
```

Comment:
```suggestion
    opts = keyword!(opts, shape: {}, type: {:f, 32}, min_val: 0.0, max_val: 1.0)
```

Reply:
Adjusted with a fix to random_bits for support of constants